### DISCUSSION TITLE
How to see all slides in overview mode?

### DISCUSSION THREAD

**Discussion by libralibra** [hakimel/reveal.js#3433] [General]
Is there an easy way to see all slides in the overview mode? Currently, only a few slides are shown in overview mode. I am trying to find a way that shows all of them like what **flowtime.js** does. Any help would be appreciated.

**Comment by t-fritsch** [comment]
There is no way to do so using the current API, Reveal does the "zooming out" effect in overview mode by setting a `transform: scale(...)` css style on the `.slides` div. This scale is calculated here : https://github.com/hakimel/reveal.js/blob/92ee97fbfe74239e8bc196b0954bd19a23410fd4/js/controllers/overview.js#L121-L122

As you can see values are hard coded and thus you're not able to configure it directly.

The good news is you can use a MutationObserver to track when Reveal changes the scale of the `.slides` div and override it with the scale you need. 

```js
// regexp to match only the last `scale(...)` rule (there are two scale rules on the .slides element, don't know why)
const currentScaleRegexp = /scale\([^\)]*\)(?!.*(scale))/;
// prevent infinite loop https://stackoverflow.com/a/48599188
let insideInitialObserverCallback = false;
    
let observer = new MutationObserver( mutationsList => {
    // only react when in overview mode
    // could also use https://revealjs.com/overview/#events to stop/start observer when needed
    if (!Reveal.isOverview()) return; 

    insideInitialObserverCallback = ! insideInitialObserverCallback
    if ( insideInitialObserverCallback ) {
        mutationsList.forEach(mutation => {
            const currentTransform = mutation.target.style.transform;
            // change to the scale that fits your needs
            mutation.target.style.transform = currentTransform.replace(currentScaleRegexp, 'scale(.05)');
        })
    }
} );
observer.observe(
    Reveal.getSlidesElement(),
    { attributes: true, attributesFilter: ['style'] }
);
```

when working with small scale like this, you may want to make the "current slide" styling more visible than the default one :
```css
.reveal.overview .slides section.present {
    outline-color: cyan; /* color that suits your theme */
    outline-width: 20px;
    outline-offset: 20px;
}
```

liimitations : 
- the overview still gets centered on the curent slide (not the same as flowtime). If you need fixed overview, you could try to override `translate()` rules too.
- the `viewDistance` gets forced to `10`, so it will work if your presentations are not too big, if they are, expect to see some slides on left/right/top/bottom ends to disapear/appear : https://github.com/hakimel/reveal.js/blob/92ee97fbfe74239e8bc196b0954bd19a23410fd4/js/reveal.js#L1751 


hope this helps 🙏

**Reply by libralibra** [reply]
@t-fritsch Thank you very much. I will give it a try.
actually, I did change the current slide style using CSS to make it stand out from the others. I also added a hover style to make it easier to navigate in overview mode using the mouse.

I have limited knowledge of JS -  can read and write small pieces of code snippets. When I tried in the first place, I cannot find where to start. Usually, the HTML file links to js and CSS. But the latest version of reveal.js links to a 'compiled' version of js (maybe?) in the dist folder, which makes it difficult for humans to read the code. When I manually changed the link to the js folder, it didn't work.

How can I link to **js/reveal.js** file? Currently, in the HTML file it points to **dist/reveal.js**. If I change it to the modified version of **js/reveal.js**, it breaks everything and no slide is shown.

**Reply by t-fritsch** [reply]
the code I provided is to be copy/pasted in your html file, in the `<script>` node, right after your `Reveal.initialize(...)` call.

Regarding the `js/reveal.js` file, you're right about it : it is a source file that gets compiled to `dist/reveal.js`. 
If you'd like to edit this file and tweak it, you should first follow **"Full setup"** installation guide : https://revealjs.com/installation/#full-setup

Then when you'll edit the `js/reveal.js` file (source) it will automatically be compiled to `dist` folder and auto-reload your presentation in the browser. 

If you're not familiar with javascript, I suggest you keep opened [the console tab of the developer tools](https://firefox-source-docs.mozilla.org/devtools-user/web_console/index.html) in your browser, so that you can quickly see if your modifications create new JS errors 😉

**Reply by libralibra** [reply]
@t-fritsch really appreciate your time and reply.

Now, I understand why the direct link to Js file didn't work. I am hoping to modify the overview toggle function directly to make it calculate the proper scale by itself. It looks like I have to go for the full-setup route. 

Would you please clarify the "automatically compile" operation? I didn't see a section in the page. Do you mean that I just need to make changes to js folder files? How to get the changes to be reflected to the dist version? Sorry for such a basic but "crucial to me" question~

**Reply by t-fritsch** [reply]
when you launch `npm start` command, it will "watch" for changes in the js/ folder and then rebuild (re-compile) into the dist folder, and then reload browser.

all this process is done in the gulpfile : https://github.com/hakimel/reveal.js/blob/92ee97fbfe74239e8bc196b0954bd19a23410fd4/gulpfile.js#L308

anyway modifying the source code is imo overkill in your situation, did you try the code I provided earlier ? simply adjusting the scale level should do the trick 😉

**Reply by libralibra** [reply]
@t-fritsch Yes, I did try your approach this morning and it did work 👍 ! The slides are smaller now.

as long as the gulpfile takes care of all the compiling process, it seems to straightforward that I only need to modify the js file and make it work. Then, the HTML doesn't need to be change since the dist folder will be updated automatically! Great!

**Reply by t-fritsch** [reply]
great ! just be aware that modifying the source code will make updating reveal for the next version more complex (you'll have to manage conflicts) 😉

**Reply by libralibra** [reply]
@t-fritsch The reveal.js is really flexible and powerful after trying the flowtime.js framework. It shares a similar 2D grid layout concept of slides but the syntax is simpler than the flowtime. 

As a presentation framework, I am quite happy with its current capabilities. I have made a local copy and may modify it for my personal use without contaminating the upstream repository.

**Reply by t-fritsch** [reply]
I find this is really a great tool too (this is why I try to help when I can ^^), I'm happy you like it too !

Regarding the potential conflicts I was not worried about upstream but about you losing your changes when you upgrade 😉
